File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Go/4418eb8ac9d0234e17ff1694fffa2258.go
package main

import (
	"fmt"
)

type Info struct {
	time [2]int64
	p    int64
}

func Merge(a, b []Info) []Info {
	ans := make([]Info, 0)
	i, j := 0, 0
	for i < len(a) && j < len(b) {
		if a[i].time[0]*b[j].time[1] == a[i].time[1]*b[j].time[0] {
			ans = append(ans, Info{a[i].time, (a[i].p + b[j].p) % mod})
			i++
			j++
		} else if a[i].time[0]*b[j].time[1] < a[i].time[1]*b[j].time[0] {
			if a[i].p != 0 {
				ans = append(ans, a[i])
			}
			i++
		} else {
			if b[j].p != 0 {
				ans = append(ans, b[j])
			}
			j++
		}
	}
	for i < len(a) {
		ans = append(ans, a[i])
		i++
	}
	for j < len(b) {
		ans = append(ans, b[j])
		j++
	}
	return ans
}

func MergeDp(a, b []Info) []Info {
	ra := make([]Info, len(a))
	rb := make([]Info, len(b))
	copy(ra, a)
	copy(rb, b)
	for i, j, sum := 0, 0, int64(0); i < len(a); i++ {
		for j < len(b) && b[j].time[0]*a[i].time[1] >= b[j].time[1]*a[i].time[0] {
			sum = (sum + b[j].p) % mod
			j++
		}
		ra[i].p = (ra[i].p * sum) % mod
	}
	for i, j, sum := 0, 0, int64(0); j < len(b); j++ {
		for i < len(a) && a[i].time[0]*b[j].time[1] > a[i].time[1]*b[j].time[0] {
			sum = (sum + a[i].p) % mod
			i++
		}
		rb[j].p = (rb[j].p * sum) % mod
	}
	return Merge(ra, rb)
}

func main() {
	var n int
	fmt.Scan(&n)
	orig := make([][3]int64, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&orig[i][0], &orig[i][1], &orig[i][2])
		orig[i][2] = (orig[i][2] * pow(100, mod-2, mod)) % mod
	}

	allInfo := make([][][4]Info, n-1)
	allSize := 0

	for l := 0; l < n-1; l++ {
		for cate := 0; cate < 4; cate++ {
			r := l + 1
			p := int64(1)
			dv := int64(0)
			if cate&1 == 1 {
				p = (p * (mod + 1 - orig[l][2])) % mod
				dv += orig[l][1]
			} else {
				p = (p * orig[l][2]) % mod
				dv -= orig[l][1]
			}
			if cate&2 == 2 {
				p = (p * (mod + 1 - orig[r][2])) % mod
				dv -= orig[r][1]
			} else {
				p = (p * orig[r][2]) % mod
				dv += orig[r][1]
			}
			dv = max(0, dv)
			allInfo[l][cate] = []Info{{[2]int64{orig[r][0] - orig[l][0], dv}, p}}
			allSize++
		}
	}

	for i := 0; (1 << i) < (n - 1); i++ {
		for j := 0; j*(1<<(i+1)) < n-1; j++ {
			noA := j * (1 << (i + 1))
			noB := noA + (1 << i)
			if noB >= n-1 {
				continue
			}
			for f1 := 0; f1 < 2; f1++ {
				for f2 := 0; f2 < 2; f2++ {
					rans1 := MergeDp(allInfo[noA][f1], allInfo[noB][f2<<1])
					p1 := pow(mod+1-orig[noB][2], mod-2, mod)
					for i := 0; i < len(rans1); i++ {
						rans1[i].p = (rans1[i].p * p1) % mod
					}
					rans2 := MergeDp(allInfo[noA][f1+2], allInfo[noB][(f2<<1)+1])
					p2 := pow(orig[noB][2], mod-2, mod)
					for i := 0; i < len(rans2); i++ {
						rans2[i].p = (rans2[i].p * p2) % mod
					}
					tempInfo := Merge(rans1, rans2)
					for i := 0; i < 4; i++ {
						allInfo[noA][i] = tempInfo
					}
				}
			}
		}
	}

	ans := int64(0)
	for cate := 0; cate < 4; cate++ {
		for _, it := range allInfo[0][cate] {
			if it.time[1] != 0 {
				ans = (ans + it.p*it.time[0]%mod*pow(it.time[1], mod-2, mod)) % mod
			}
		}
	}
	fmt.Println(ans)
}

const (
	maxn = 100100
	mod  = 998244353
)

func pow(n, m, mod int64) int64 {
	if m < 0 {
		return 0
	}
	ans := int64(1)
	k := n
	for m > 0 {
		if m&1 == 1 {
			ans = (ans * k) % mod
		}
		k = (k * k) % mod
		m >>= 1
	}
	return ans
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/4418eb8ac9d0234e17ff1694fffa2258.c
#include <stdio.h>
#include <stdlib.h>

#define MOD 998244353

typedef long long ll;

typedef struct {
    ll first;
    ll second;
} pll;

typedef struct {
    pll time;
    ll p;
} Info;

int n;
pll orig[100100];

Info all_info[100100][4];
Info temp_info[4];
int all_size = 0;

ll pow(ll n, ll m, ll mod) {
    if (m < 0) return 0;
    ll ans = 1;
    ll k = n;
    while (m) {
        if (m & 1) {
            ans *= k;
            if (mod) ans %= mod;
        }
        k *= k;
        if (mod) k %= mod;
        m >>= 1;
    }
    return ans;
}

void add(ll* a, ll b, int mod) {
    if (mod == -1) mod = MOD;
    *a += b;
    while (*a >= mod) *a -= mod;
    while (*a < 0) *a += mod;
}

int jud(double a, double b) {
    if (abs(a) < 1e-10 && abs(b) < 1e-10)
        return 0;
    else if (abs(a - b) / max(abs(a), abs(b)) < 1e-10)
        return 0;
    if (a < b) return -1;
    return 1;
}

int find(ll val, ll* a, int na, int f_small, int f_lb) {
    if (na == 0) return 0;
    int be = 0, en = na - 1;
    if (*a <= *(a + na - 1)) {
        if (f_lb == 0)
            while (be < en) {
                int mid = (be + en + 1) / 2;
                if (jud(*(a + mid), val) != 1)
                    be = mid;
                else
                    en = mid - 1;
            }
        else
            while (be < en) {
                int mid = (be + en) / 2;
                if (jud(*(a + mid), val) != -1)
                    en = mid;
                else
                    be = mid + 1;
            }
        if (f_small && jud(*(a + be), val) == 1) be--;
        if (!f_small && jud(*(a + be), val) == -1) be++;
    } else {
        if (f_lb)
            while (be < en) {
                int mid = (be + en + 1) / 2;
                if (jud(*(a + mid), val) != -1)
                    be = mid;
                else
                    en = mid - 1;
            }
        else
            while (be < en) {
                int mid = (be + en) / 2;
                if (jud(*(a + mid), val) != 1)
                    en = mid;
                else
                    be = mid + 1;
            }
        if (!f_small && jud(*(a + be), val) == -1) be--;
        if (f_small && jud(*(a + be), val) == 1) be++;
    }
    return be;
}

int bitnum(unsigned int nValue) { return __builtin_popcount(nValue); }
int bitnum(int nValue) { return __builtin_popcount(nValue); }
int bitnum(unsigned long long nValue) {
    return __builtin_popcount(nValue) + __builtin_popcount(nValue >> 32);
}
int bitnum(ll nValue) {
    return __builtin_popcount(nValue) + __builtin_popcount(nValue >> 32);
}
int id0(unsigned int a) {
    if (a == 0) return 0;
    return 32 - __builtin_clz(a);
}
int id0(int a) {
    if (a == 0) return 0;
    return 32 - __builtin_clz(a);
}
int id0(unsigned long long a) {
    int temp = a >> 32;
    if (temp) return 32 - __builtin_clz(temp) + 32;
    return id0((int)a);
}
int id0(ll a) {
    int temp = a >> 32;
    if (temp) return 32 - __builtin_clz(temp) + 32;
    return id0((int)a);
}

ll min1(ll a, ll b) {
    return a < b ? a : b;
}

ll max1(ll a, ll b) {
    return b < a ? a : b;
}

Info* Merge(Info* a, int na, Info* b, int nb) {
    Info* ans = (Info*)malloc((na + nb) * sizeof(Info));
    int i = 0, j = 0, k = 0;
    for (; i < na && j < nb; ) {
        if (a[i].time.first * b[j].time.second == a[i].time.second * b[j].time.first) {
            ans[k].time = a[i].time;
            ans[k].p = (a[i].p + b[j].p) % MOD;
            i++;
            j++;
            k++;
        } else if (a[i].time.first * b[j].time.second < a[i].time.second * b[j].time.first) {
            if (a[i].p) {
                ans[k].time = a[i].time;
                ans[k].p = a[i].p;
                k++;
            }
            i++;
        } else {
            if (b[j].p) {
                ans[k].time = b[j].time;
                ans[k].p = b[j].p;
                k++;
            }
            j++;
        }
    }
    while (i < na) {
        ans[k].time = a[i].time;
        ans[k].p = a[i].p;
        i++;
        k++;
    }
    while (j < nb) {
        ans[k].time = b[j].time;
        ans[k].p = b[j].p;
        j++;
        k++;
    }
    return ans;
}

Info* MergeDp(Info* a, int na, Info* b, int nb) {
    Info* ra = a;
    Info* rb = b;
    for (ll i = 0, j = 0, sum = 0; i < na; i++) {
        while (j < nb && b[j].time.first * a[i].time.second >= b[j].time.second * a[i].time.first) {
            add(&sum, b[j].p, MOD);
            j++;
        }
        ra[i].p *= sum;
        ra[i].p %= MOD;
    }
    for (ll i = 0, j = 0, sum = 0; j < nb; j++) {
        while (i < na && a[i].time.first * b[j].time.second > a[i].time.second * b[j].time.first) {
            add(&sum, a[i].p, MOD);
            i++;
        }
        rb[j].p *= sum;
        rb[j].p %= MOD;
    }
    return Merge(ra, na, rb, nb);
}

int main() {
    freopen("input.txt", "r", stdin);
    scanf("%d", &n);
    for (int i = 0; i < n; i++) {
        scanf("%lld%lld%lld", &orig[i].first, &orig[i].second.first, &orig[i].second.second);
        orig[i].second.second *= pow(100, MOD - 2, MOD);
        orig[i].second.second %= MOD;
    }

    for (int l = 0; l < n - 1; l++) {
        for (int cate = 0; cate < 4; cate++) {
            int r = l + 1;
            ll p = 1, dv = 0;
            p *= (cate & 1) ? orig[l].second.second : MOD + 1 - orig[l].second.second;
            dv += (cate & 1) ? orig[l].second.first : -orig[l].second.first;
            p *= (cate & 2) ? orig[r].second.second : MOD + 1 - orig[r].second.second;
            dv += (cate & 2) ? -orig[r].second.first : orig[r].second.first;
            dv = max1(0ll, dv);
            p %= MOD;
            all_info[l][cate].time.first = orig[r].first - orig[l].first;
            all_info[l][cate].time.second = dv;
            all_info[l][cate].p = p;
            all_size++;
        }
    }
    for (int i = 0; (1 << i) < (n - 1); i++) {
        for (int j = 0; j * (1 << (i + 1)) < n - 1; j++) {
            int no_a = j * (1 << (i + 1)), no_b = no_a + (1 << i);
            if (no_b >= n - 1) continue;
            for (int f1 = 0; f1 < 2; f1++) {
                for (int f2 = 0; f2 < 2; f2++) {
                    Info* rans1 = MergeDp(all_info[no_a] + f1, 2, all_info[no_b] + (f2 << 1), 2);
                    ll p1 = pow(MOD + 1 - orig[no_b].second.second, MOD - 2, MOD);
                    for (int k = 0; k < 2; k++) {
                        rans1[k].p *= p1;
                        rans1[k].p %= MOD;
                    }
                    Info* rans2 = MergeDp(all_info[no_a] + f1 + 2, 2, all_info[no_b] + (f2 << 1) + 1, 2);
                    ll p2 = pow(orig[no_b].second.second, MOD - 2, MOD);
                    for (int k = 0; k < 2; k++) {
                        rans2[k].p *= p2;
                        rans2[k].p %= MOD;
                    }
                    temp_info[f1 + (f2 << 1)] = Merge(rans1, 2, rans2, 2);
                }
            }
            for (int i = 0; i < 4; i++) {
                all_info[no_a][i] = temp_info[i];
            }
        }
    }

    ll ans = 0;
    for (int cate = 0; cate < 4; cate++) {
        for (int i = 0; i < 2; i++) {
            if (all_info[0][cate].time.second) {
                add(&ans, all_info[0][cate].p * all_info[0][cate].time.first % MOD * pow(all_info[0][cate].time.second, MOD - 2, MOD), MOD);
            }
        }
    }
    printf("%lld\n", ans);

    return 0;
}
